### PR TITLE
Adding rsync -y option to compliment -x

### PR DESCRIPTION
**Pull request by thomasmaclean**

### CODE REVIEW

File: gslib/tests/test_rsync.py [hunk 1]
```diff
@@ -2733,7 +2733,7 @@ def test_dir_to_bucket_minus_x(self):
     @Retry(AssertionError, tries=3, timeout_secs=1)
     def _Check1():
       """Tests rsync works as expected."""
-      self.RunGsUtil(['rsync', '-d', '-x', 'obj[34]', tmpdir, suri(bucket_uri)])
+      self.RunGsUtil(['rsync', '-d', '-y', 'obj[34]', tmpdir, suri(bucket_uri)])
```
Comment (dilipped) [inline gslib/tests/test_rsync.py:2736]: should this be done in a loop for -x as well? Like you did it for another test down below. Ideally we would use parameterized, but I am guessing gsutil does not have that library.

**Review by dilipped — APPROVED** [review]

File: gslib/commands/rsync.py [hunk 2]
```diff
@@ -528,6 +528,24 @@
                  use ^ as an escape character instead of \\ and escape the |
                  character. When using Windows PowerShell, use ' instead of "
                  and surround the | character with ".
+
+  -y pattern     Similar to the -x option, but the command will first skip 
+                 directories/prefixes using the provided pattern and then
+                 exclude files/objects using the same pattern. This is usually
+                 much faster, but will fail to work with with negative
```
[the remark this answers is not in the document]
Comment (thomasmaclean) [reply gslib/commands/rsync.py]: I rewrote this whole section to make it a bit more clear about what the resulting behavior will be.

File: gslib/wildcard_iterator.py [hunk 3]
```diff
@@ -758,7 +758,9 @@ def _ExcludeDir(self, dir):
     """
```
[the remark this answers is not in the document]
Comment (thomasmaclean) [reply gslib/wildcard_iterator.py:758]: Done.

File: gslib/tests/test_rsync.py [hunk 4]
```diff
@@ -2757,6 +2757,38 @@ def _Check2():
...
+        set([
+            '/a', '/b', '/c', '/data1/a.txt', '/data1/ok', '/data2/b.txt',
+            '/data3/data4/c.txt'
+        ]))
+    stderr = self.RunGsUtil(['ls', suri(bucket_uri, '**')],
+                            expected_status=1,
+                            return_stderr=True)
+    self.assertIn('One or more URLs matched no objects', stderr)
+    self.RunGsUtil(
+        ['rsync', '-r', '-x', '^(?!.*\.txt$).*', tmpdir,
```
[the remark this answers is not in the document]
Comment (thomasmaclean) [reply gslib/tests/test_rsync.py]: I rewrote these to be pseudo-parameterized.

File: gslib/tests/test_rsync.py (hunk 1 above)
Comment (thomasmaclean) [reply gslib/tests/test_rsync.py]: I rewrote these to be pseudo-parameterized.

**Pull request merged**